ivre/db/mongo.py

```diff
@@ -1142,6 +1142,32 @@ def searchscreenshot(port=None, protocol='tcp', service=None, neg=False):
                            'screenshot': {'$exists': not neg}}
         }}
 
+    @staticmethod
+    def searchcpe(value=None, type=None, vendor=None, product=None,
+                  components=None):
+        """Look for a CPE by value (original cpe string), type (a, o or h),
+        vendor, product or components (the part after the column following
+        the product). No argument will just check for cpe existence.
+
+        """
+        if all(arg is None
+               for arg in [value, type, vendor, product, components]):
+            return {"cpes": {"$exists": True}}
+
+        flt = {}
+        fields = [
+            ("value", value),
+            ("type", type),
+            ("vendor", vendor),
+            ("product", product),
+            ("components", components),
+        ]
+
+        for (field_name, field_val) in fields:
+            if field_val is not None:
+                flt["cpes." + field_name] = field_val
+        return flt
```

Comment:
You are right (I'm not very familiar with the MongoDB API yet). What would you think of:
```python
        fields = [
            ("value", value),
            ("type", cpe_type),
            ("vendor", vendor),
            ("product", product),
            ("components", components),
        ]
        flt = dict((field, value) for field, value in fields
                                  if value is not None)
        nflt = len(flt)
        if nflt == 0:
            return {"cpes": {"$exists": True}}
        else:
            return {"cpes": {"$elemMatch": flt}}
```
No special case for 1 element match, but shorter code.